integration_test/integration-test.go

```diff
@@ -0,0 +1,324 @@
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+/*****************************
+* INTEGRATION TEST FRAMEWORK *
+******************************/
+
+type Map = map[string]string
+
+/*
+* TestSuite struct
+*   keep track of info used by a test suite
+*   different proxy configurations corresponds
+*   to different test suites
+ */
+type TestSuite struct {
+	// url of the proxy to be tested
+	// the test client will send requests to this url
+	proxyURL string
+
+	// the port that the test server will be listening at
+	serverPort string
+
+	// the expected message received from the test server if
+	// the request has been successful
+	serverMessage string
+
+	// the server that will be responding requests from the proxy
+	server http.Server
+
+	// the channels for communication between the server
+	// and the client. Received requests will be pushed through these
+	// channels by the server for the client to verify
+	requestReceived    chan *http.Request
+	bodyStringReceived chan string
+
+	// a collection of requests to run in this suite
+	// each test has a test name associated with it
+	requests map[string]TestRequest
+}
+
+/*
+* TestRequest struct
+*   keep track of info used by one request in a test suite
+ */
+type TestRequest struct {
+	// the test request to be made
+	path       string
+	pathParams url.Values
+	headers    Map
+	cookies    Map
+	body       url.Values
+
+	// the expected sqli detection result for the proxy
+	// if true, the request contains sqli
+	sqli      bool
+	sqli_body bool
+}
+
+/*
+* Initiate a new test suite with a given proxyURL
+ */
+func NewTestSuite(proxyURL string) TestSuite {
+	return TestSuite{
+		proxyURL:           proxyURL,
+		serverPort:         ":8080",
+		serverMessage:      "request received",
+		requestReceived:    make(chan *http.Request),
+		bodyStringReceived: make(chan string),
+		requests:           map[string]TestRequest{},
+	}
+}
+
+/*
+* Run all requests in a test suite
+ */
+func (suite TestSuite) RunRequests() {
+	suite.StartServer()
+	suite.CheckProxyConnection()
+	for name, test := range suite.requests {
+		suite.RunTestRequest(name, test)
+	}
+	suite.CloseServer()
+}
+
+/*
+* Initiate a new test request with default content
+ */
+func NewTestRequest() TestRequest {
+	return TestRequest{
+		path:       "/",
+		pathParams: url.Values{"path-key-0": {"path-val-0"}},
+		headers:    Map{"header-key-0": "header-val-0"},
+		cookies:    Map{"cookie-key-0": "cookie-val-0"},
+		body:       url.Values{"body-key-0": {"body-val-0"}},
+		sqli:       false,
+		sqli_body:  false,
+	}
+}
+
+/*
+* Run a given test in a test suite
+* assuming the server and the proxy has been running
+ */
+func (suite TestSuite) RunTestRequest(testName string, test TestRequest) {
+	result := suite.Eval(test)
+	if !result {
+		fmt.Println("TEST FAILED: " + testName)
+		fail()
+	}
+	fmt.Println("TEST PASSED: " + testName)
+}
+
+/*
+* Initialize and Start the test server
+ */
+func (suite TestSuite) StartServer() {
```

Comment:
do we need a separate test server? It seems it could just be supported with either of the following optoins:
1. configure Envoy filter to echo back the request as-is, no extra server needed
2. reuse the common httpbin docker image which is standard tool used in many tests

either way we don't need to write our own server unless you really need it for cases that cannot be supported by above options.

Reply:
Thanks for the suggestion! It sounds like both options may suffice the needs for integration test of this module. 

For option 1, I am currently not sure how to configure Envoy to do this. I will look up the documentations.
For option 2, I haven't heard of the httpbin docker image before. I will also look it up. Also, are there any particular integration tests that I should read as a reference? Before my implementation of the tests, I read the envoye2e test for istio extension stats plugin [here](https://github.com/istio/proxy/tree/master/test/envoye2e/stats_plugin) by another intern Vincente Ferrara, and the ABI layer test framework in proxy-wasm [here](https://github.com/proxy-wasm/test-framework) still under development by another intern Chris Agia.